network/transport/v2/protocol.go

```diff
@@ -187,13 +189,28 @@ func (p *protocol) Configure(_ transport.PeerID) error {
 func (p *protocol) Start() (err error) {
 	p.cMan = newConversationManager(maxValidity)
 	p.cMan.start(p.ctx)
+	p.routines = new(sync.WaitGroup)
 
 	if p.config.DiagnosticsInterval > 0 {
-		p.diagnosticsMan.start(p.ctx, time.Duration(p.config.DiagnosticsInterval)*time.Millisecond)
+		p.routines.Add(1)
+		go func(w *sync.WaitGroup) {
+			defer w.Done()
+			p.diagnosticsMan.start(p.ctx, time.Duration(p.config.DiagnosticsInterval)*time.Millisecond)
+		}(p.routines)
 	}
 
-	p.listHandler = newTransactionListHandler(p.ctx, p.handleTransactionList)
-	p.listHandler.start()
+	ctx := p.ctx // use copy of pointer to avoid data race
+	// Wrap listHandler function to supply the context to handleTransactionList.
+	// It would be prettier to pass a context around in all protocol message handlers (since most use context.Background() for database access now),
+	// but that is too big a change for now.
+	p.listHandler = newTransactionListHandler(p.ctx, func(peer transport.Peer, envelope *Envelope) error {
```

Comment:
change the `handleFunc` signature so the ctx is passed as well and you don't need a copy. HandleTransactionList should reuse the context passed by the TransctionListHandler, not the context from the protocol.